### PR TITLE
[narwhal] fix executor recovery 

### PR DESCRIPTION
**Pull request by akichidis**
During a crash recovery we are sending to the executor all the certificates that `have been committed` but `not executed` yet. To achieve that both the `consensus_store` and `state` are being used to determine what's the latest point we executed a certificate and what is latest committed certificate. 

On a recent incident we observed that during the recovery of the certificates two things happened:
1. Instead of starting the "recovery" of the last executed certificate, we did start one certificate before. Ex instead of start (re)sending the committed certificate from index `308807` , we did practically start sending the certificate from index `308806`.
2. The associated index number and certificate were mismatched. So although we did sent the certificate from round `308807` , we assigned it the index `308808` , which is wrong
3. Also, based on the code, we were effectively skipping sending the last committed certificate. So effectively the transactions from that certificate would never get executed.

This PR is fixing the above issues by ensuring that:
* during recovery we are starting (re)sending the unexecuted certificates **from the right point** & **use the right index**
* we are always resending all the unexecuted certificates without omitting the last one

**Note:** Also took the opportunity and enhanced a little bit the logging on a couple of more places.

### CODE REVIEW

File: narwhal/executor/src/lib.rs
Comment (akichidis) [inline narwhal/executor/src/lib.rs:122]: Added this ticket as place holder. I would like us to refactor the `next_certificate_index` semantics as it's super confusing the fact that we always store alongside not the index of the certificate that we just committed, but the "next" one.

File: narwhal/executor/src/lib.rs
Comment (andll) [inline narwhal/executor/src/lib.rs:147]: Maybe cleaner option is to just change `read_sequenced_certificates(range)` to `read_sequenced_certificates_from(start: u64)` and just read the table until the end?
This way we won't need `consensus_next_index` variable at all and we avoid confusion about what it means

The `if last_executed_index <= consensus_next_index` could also be deleted in that case

**Review by andll — COMMENTED** [review]
Good find @akichidis

I see that main confusion seem to come from `bullshark.rs` where we send `ConsensusOutput{consensus_index, ...}` and then increment `consensus_index` and then write **incremented** sequence to the database.

I wonder if this would be an actual place to fix the problem, since it is incredibly strange to send one index over channel but put something else in the DB, wdyt?

I also added few comments to code about how I think we can remove more confusing pieces

Overall I think PR fixes the issue, but since this type of confusion is very concerning I am curious what do you think about suggestions above?

File: narwhal/executor/src/lib.rs
Comment (laura-makdah) [inline narwhal/executor/src/lib.rs:135]: why the +1?

File: narwhal/executor/src/lib.rs
Comment (laura-makdah) [inline narwhal/executor/src/lib.rs:149]: why do we decrement the sequence?

File: narwhal/executor/src/lib.rs
Comment (akichidis) [reply narwhal/executor/src/lib.rs:147]: Good point @andll , can certainly do like this

**Comment by akichidis** [comment]
> Good find @akichidis
> 
> I see that main confusion seem to come from `bullshark.rs` where we send `ConsensusOutput{consensus_index, ...}` and then increment `consensus_index` and then write **incremented** sequence to the database.
> 
> I wonder if this would be an actual place to fix the problem, since it is incredibly strange to send one index over channel but put something else in the DB, wdyt?
> 
> I also added few comments to code about how I think we can remove more confusing pieces
> 
> Overall I think PR fixes the issue, but since this type of confusion is very concerning I am curious what do you think about suggestions above?

@andll I totally agree with the suggestion to make the change in `bullshark.rs` , and basically that's why I initially put this comment here https://github.com/MystenLabs/sui/pull/5820#discussion_r1013010568 . Ultimately I would like to refactor even the semantics of "next" etc which are super confusing to the way we use them today. That being said as part of this PR I was looking to not touch core parts of `bullshark` and db tables , and do that as part of upcoming PR. Sounds good/reasonable to you or do you still see more value just going with it now?

**Comment by andll** [comment]
> That being said as part of this PR I was looking to not touch core parts of bullshark and db tables , and do that as part of upcoming PR

I will leave it up to you then, I *thought* that updating bullshark.rs will be roughly same number of LoC as this PR, but I have not looked in details how many things this will touch. If you think it is more contained to do the fix the way current PR is structured this sounds good to me

**Review by andll — APPROVED** [review]
Accepting but let's do `read_sequenced_certificates_from` fix before landing

File: crates/sui-core/src/consensus_handler.rs
Comment (akichidis) [reply crates/sui-core/src/consensus_handler.rs:73]: The reason I am changing this is because with the previous way we were overwriting the `consensus_index` value and assigning the value `None` when `Self::update_hash` returned such. That makes the debug log in line 82 print `Ignore consensus transaction at index None....`

File: narwhal/consensus/src/bullshark.rs
Comment (mwtian) [inline narwhal/consensus/src/bullshark.rs:180]: small nit: do we need to log both &consensus_index - 1 and &consensus_index?

**Review by mwtian — APPROVED** [review]

File: narwhal/executor/src/lib.rs
Comment (akichidis) [reply narwhal/executor/src/lib.rs:149]: Exactly. The returned indexes from the `read_sequenced_certificates` are ahead +1 . So in order to align them to what we are sending to the executor we should do the decrement

**Pull request merged**
